cmd/server/main.go

```diff
@@ -62,6 +65,210 @@ type BindingSecret struct {
+	if err != nil {
+		panic(err.Error())
+	}
+
+	clientsOrDie(config)
+
+	log.Print("Entering watch loop")
+
+	go startPollingUPS()
+	startWatchLoop()
+}
+
+func startWatchLoop() {
+	events, err := k8client.CoreV1().Secrets(os.Getenv(NamespaceKey)).Watch(metav1.ListOptions{})
+	if err != nil {
+		panic(err.Error())
+	}
+
+	for update := range events.ResultChan() {
+		switch action := update.Type; action {
+		case ActionAdded:
+			handleAddSecret(update.Object)
+		case ActionDeleted:
+			handleDeleteSecret(update.Object)
+		default:
+			log.Print("Unhandled action:", action)
+		}
+	}
+}
+
+func handleAddSecret(obj runtime.Object) {
+	raw, _ := json.Marshal(obj)
+	var secret = BindingSecret{}
+	json.Unmarshal(raw, &secret)
+	if val, ok := secret.Labels[SecretTypeKey]; ok && val == BindingSecretType {
+		appType := string(secret.Data[BindingAppType])
+		log.Printf("A mobile binding secret of type `%s` was added", appType)
+
+		if appType == "Android" {
+			handleAndroidVariant(&secret)
+		} else if appType == "IOS" {
+			handleIOSVariant(&secret)
+		}
+		// Always delete the secret after handling it regardless of any new resources
+		// was created
+		deleteSecret(secret.Name)
+	}
+}
+
+func handleDeleteSecret(obj runtime.Object) {
+	raw, _ := json.Marshal(obj)
+	var secret = BindingSecret{}
+	json.Unmarshal(raw, &secret)
+
+	for _, ref := range secret.ObjectMeta.OwnerReferences {
+		if ref.Kind == "ServiceBinding" {
+			handleDeleteVariant(&secret)
+			break
+		}
+	}
+}
+
+// startPollingUPS() is a loop that calls comparseUPSVariantsWithClientConfigs() in intervals
+func startPollingUPS() {
+	interval := UPSPollingInterval * time.Second
+	for {
+		<-time.After(interval)
+		compareUPSVariantsWithClientConfigs()
+	}
+}
+
+// compareUPSVariantsWithClientConfigs() compares the UPS client configs stored in k8's secrets
+// against the variants in UPS in order to detect if a variant has been deleted in UPS
+// If a client config is found that references a variant not found in UPS then we clean up the client config by deleting the associated servicebinding.
+func compareUPSVariantsWithClientConfigs() {
+
+	err := initPushClient()
+
+	if err != nil {
+		log.Printf("error initialising UPS client: %s", err.Error())
+		return
+	}
+
+	// get the UPS related secrets
+	secrets, err := getUPSSecrets()
+
+	if err != nil {
+		log.Printf("Error searching for ups secrets: %v", err.Error())
+		return
+	}
+
+	// process the secrets into a list of VariantServiceBindingMappings
+	// each element has VariantId and ServiceBindingId
+	clientConfigs := getUPSVariantServiceBindingMappings(secrets)
+
+	// Get all variants from UPS
+	UPSVariants, err := pushClient.getVariants()
+
+	if err != nil {
+		log.Printf("An error occurred trying to get variants from UPS service: %v", err.Error())
+		return
+	}
+
+	for _, clientConfig := range clientConfigs {
+		found := false
+
+		for _, variant := range UPSVariants {
+			if variant.VariantID == clientConfig.VariantId {
+				found = true
```

Comment:
We could exit the loop early here, no reason to inspect the other variants when we already found the matching one